proposals/0000-decl-invis-binders.rst

```diff
@@ -0,0 +1,440 @@
+inference would include as few forall-bound variables as possible, to allow
+higher-kinded usage. However, it shall also include all ``@``-bound variables::
+
+  -- arity = 1
+  type F :: Type -> forall k. Maybe k
+  type family F x
+
+  -- arity = 2
+  type G :: Type -> forall k. Maybe k
+  type family G x @k
+
+This would simultaneously reduce code duplication and simplify the rules for
+arity inference.
+
+Motivation Summary
+~~~~~~~~~~~~~~~~~~
+
+Support for ``@k``-binders in type declarations will have the following
+positive effects:
+
+* Term- and type-level syntax become more consistent with one another
+* ``class`` and ``instance`` declarations can be more symmetric
+* Standalone kind signatures will no longer lead to duplication
+* Implicit quantification rules in type synonyms can be simplified
+* Arity inference becomes much more straightforward
+
+Proposed Change Specification
+-----------------------------
+
+Background: SAKS Zipping
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+**SAKS zipping** is the part of checking declarations against the corresponding
+standalone kind signature that pairs quantifiers with zero-or-one binders.
+
+Consider the following declaration::
+
+  type T :: forall a. a -> forall b c. (b, c) -> forall d -> (a ~ b) => Type
+  data T (x :: Type) y z = ...
+
+Here we produce the following pairs::
+
+    Quantifier  |   Binder
+  --------------+------------
+    forall a.   |
+    a ->        | (x :: Type)
+    forall b.   |
+    forall c.   |
+    (b, c) ->   | y
+    forall d -> | z
+    (a ~ b) =>  |
+
+SAKS zipping works over two lists: quantifiers and binders. Let us define it in
+pseudo-code::
+
+  -- Base version (before this proposal)
+  zipSAKS (q:qs) (b:bs)
+    | isInvisibleQuantifier q = (q, Nothing) : zipSAKS qs (b:bs)
+    | otherwise               = (q, Just b)  : zipSAKS qs bs
+  zipSAKS [] (b:bs) = error "Too many binders"
+  zipSAKS _ [] = []
+
+``isInvisibleQuantifier`` holds for ``forall a.``, ``forall {a}.``, and ``ctx
+=>``, but does not hold for ``a ->`` or ``forall x ->``.
+
+Background: Trailing Quantifiers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When SAKS zipping is done, some quantifiers may remain. Consider::
+
+  type T :: forall a. Type -> forall b. Type
+  type T x = ...
+
+The produced pairs are::
+
+    Quantifier  |   Binder
+  --------------+------------
+    forall a.   |
+    Type ->     | x
+
+Zipping stops when binders are exhausted, so the ``forall b.`` does not yield a
+pair. Instead, it becomes a part of the return type. We call the remaining
+quantifiers **trailing**.
+
+During arity inference, it is determined how many of the trailing ``forall``
+quantifiers are included in the arity of a type synonym or type family.
+
+Primary Change: ``@k``-binders
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Relax the syntactic check of ``data``, ``newtype``, ``type``, ``class``,
+   ``type family``, and ``data family`` declarations to allow ``@k``-binders in
+   their headers::
+
+     tv_bndr ::=
```

Comment:
This could be misconstrued. What is `v`? What is `k`? Is `@_` one lexeme or two? I recommend expanding this out a bit, perhaps to:

```
tv_bndr ::=
        | tyvar
        | '(' tyvar '::' kind ')'
        | '@' tyvar
        | '@' '(' tyvar '::' kind ')'
        | '@' '_'
```

With your helpful comments, and with a note saying that all `@`-signs are prefix occurrences.